messages-list-directive.js

```diff
@@ -19,42 +19,148 @@ function factory(brMessagesService, $location) {
   function Link(scope) {
     var model = scope.model = {};
     model.messages = null;
-
+    model.loading = true;
     model.sorting = {
-      subject: '+',
       date: '-'
     };
-    model.orderBy = ['-content.date', '+_id'];
+    model.orderBy = ['-date', '+id'];
+    model.showDeleteSuccessAlert = false;
+    model.showDeleteFailAlert = false;
+    if(brMessagesService.returningFrom === 'archive') {
+      model.inboxActive = false;
+      model.archiveActive = true;
+      brMessagesService.returningFrom = null;
+    } else {
+      model.inboxActive = true;
+      model.archiveActive = false;
+    }
+
+    var allMessages = [];
+    processMessageLists();
+
+    // Update the message service with the list's current URL so we can path back to it later
+    brMessagesService.setMessageListUrl($location.path());
 
     brMessagesService.getAll()
       .then(function(result) {
-        model.messages = result.data;
+        allMessages = result.data;
+        for(var message in allMessages) {
+          allMessages[message].selected = false;
+        }
+        processMessageLists();
+        model.loading = false;
         scope.$apply();
       });
 
     model.view = function(id) {
+      // TODO: $location doesn't do a full page refresh, which makes the page look a little weird
+      // when moving over because it rapidly shrinks and expands from the loading html
       $location.url('/messages/' + id);
     };
 
     model.sortClick = function(field) {
       switch(field) {
-        case 'subject':
-          model.sorting.subject = (model.sorting.subject === '+') ? '-' : '+';
-          model.orderBy = [
-            model.sorting.subject + '_id',
-            model.sorting.date + 'content.date'
-          ];
-          break;
         case 'date':
           model.sorting.date = (model.sorting.date === '+') ? '-' : '+';
           model.orderBy = [
-            model.sorting.date + 'content.date',
-            model.sorting.subject + '_id'
+            model.sorting.date + 'date'
           ];
           break;
       }
     };
 
+    model.inboxTab = function() {
+      model.inboxActive = true;
+      model.archiveActive = false;
+      model.showDeleteSuccessAlert = false;
+      model.showDeleteFailAlert = false;
+      for(var message in allMessages) {
+        allMessages[message].selected = false;
+      }
+      processMessageLists();
+    };
+
+    model.archiveTab = function() {
+      model.inboxActive = false;
+      model.archiveActive = true;
+      model.showDeleteSuccessAlert = false;
+      model.showDeleteFailAlert = false;
+      for(var message in allMessages) {
```

Comment:
Don't use `for(var x in y)` for arrays. Use `angular.forEach` or `for(var i = 0; i < x.length; ++i) {`.